Global annotations

damianhxy opened the pull request:
## Description
- Allow instructors to create, edit, and delete global annotations
- Display all non-autograded problems in annotation pane to facilitate adding global annotations
- Show a dash if a problem is released with no score (via annotation) associated with it. This avoids an internal server error when we try to format `nil`.
- Update annotations documentation (re: shared comments, global annotations, positive grading)
- Fix off-by-one line number on `viewFeedback` page
- Improve validation for max score upon problem creation
- Update logic for `problemReleased` to additionally check that the grading deadline has not passed
- Automatically destroy annotations when a problem is destroyed

## Motivation and Context
To provide instructors with a method of awarding points for a problem globally, without tying it to a specific line of code.

## How Has This Been Tested?
Test using an autograded single file assessment (e.g. hello) and a non-autograded tar assessment
- Ensure that all non-autograded problems are shown in the annotation pane
- Ensure that you can no longer create a problem without a max score
- Ensure that annotation line number shown on viewFeedback page is no longer off-by-one
- Check documentation for annotations
- Test that clickable chevron toggles collapse of a problem
- Successfully create, edit, and delete global annotations
  - Global annotations only appear in the annotation pane, and not in the code viewer
  - Global annotations still influence a student's score correctly
  - Global annotations are not clickable
- Test that above features work even after switching files or adding / editing / deleting annotations
- Ensure that students are unable to view the annotations when either (1) grades are not released and/or (2) the grading deadline has not passed
  - Ensure that students can view annotation otherwise
- Ensure that the view feedback page handles global annotations correctly
  - Clickable link loads the student's submission

**Creating a global annotation**
![Screenshot 2022-11-27 at 16 53 52](https://user-images.githubusercontent.com/9074856/204161657-0c7022c7-8248-45ee-8aac-5a4af5d23f11.png)

![Screenshot 2022-11-27 at 16 54 45](https://user-images.githubusercontent.com/9074856/204161697-f6ca577c-bf59-401e-848c-281b8136073a.png)

**Editing a global annotation**
![Screenshot 2022-11-27 at 16 54 08](https://user-images.githubusercontent.com/9074856/204161664-88150177-537f-477c-810b-ccf8e59f601e.png)

![Screenshot 2022-11-27 at 16 54 52](https://user-images.githubusercontent.com/9074856/204161702-72bbeb0c-86b2-484d-bbbb-fba4121be622.png)


**View feedback page with global annotations** - clicking the link loads the first file of the submission
![Screen Shot 2022-11-02 at 17 49 33](https://user-images.githubusercontent.com/9074856/199609163-a524c2f7-c440-4acd-a3f1-4434e1f0342a.png)

**Student unable to view annotations when unreleased**
![Screen Shot 2022-11-02 at 17 53 30](https://user-images.githubusercontent.com/9074856/199609713-250040fd-b042-4d05-a08b-2b2dbe24a089.png)

![Screen Shot 2022-11-02 at 17 54 25](https://user-images.githubusercontent.com/9074856/199609844-dc0c3373-3b5e-40bf-bc29-65f463b02b8d.png)


## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [x] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [x] I have updated the documentation accordingly, included in this PR

## Other issues / help required
- When annotations are edited, the textarea is initially too small, but they expand correctly once you start typing. I'm not sure if there's a way to make them have the right height from the start.
- The exception to this is when the file containing the annotation is directly loaded and there have been no actions (e.g. switching files, deleting another annotation, editing another annotation...). In this case, the textarea has a slightly too large height. However, the height does not decrease past the original height when deleting lines.

victorhuangwq commented on docs/features/annotations.md:
```diff
@@ -2,28 +2,52 @@
+
+## Scoring Behavior
+
+There are two intended ways for course instructors to use the add annotation features. Deductions from maximum ("negative grading"), or additions from zero ("positive grading").
```
nit: additions to zero

victorhuangwq commented on docs/features/annotations.md:
```diff
@@ -2,28 +2,52 @@
+Set a `max_score` either programmatically, or under `Edit Assessment > Problems` for the particular non-autograded question. Then when the grader is viewing the code, add a positive score, such as `5` to the score field, to add to the score. This use case is preferred when giving out bonus points.
 
-Set a `max_score` either programmatically, or under `Edit Assessment > Problems` for the particular non-autograded question to `0`. When the grader is viewing the code, add positive scores, such as `5` to the score field, to add to the score. This use case is preferred when giving out bonus points.
+The most common use case is when `max_score` is set to `0`, with marks awarded for each rubric item achieved.
```
This line is incorrect. I think with the new positive grading feature, they should just set it to whatever their max grade should be.

Perhaps you could just remove this line?

victorhuangwq reviewed (changes requested):
Shared comment aspect looks about right. Some changes requested for documentation

michellexliu commented on app/assets/javascripts/annotations.js:
```diff
@@ -1060,14 +1186,15 @@ var submitNewPDFAnnotation = function (comment, value, problem_id, pageInd, xRat
   newAnnotation.problem_id = problem_id;
   newAnnotation.filename = fileNameStr;
   newAnnotation.shared_comment = shared_comment;
+  newAnnotation.global_comment = global_comment;
```
nit: this section of all these `newAnnotation.comment = comment`, etc. feels like a lot of unnecessary code; could we just do something like
```
newAnnotation = {...newAnnotation, line: parseInt(lineInd), comment, ...,global_comment}
```

michellexliu commented on app/controllers/submissions_controller.rb:
```diff
@@ -436,7 +436,8 @@ def view
 
-    @problemReleased = @submission.scores.pluck(:released).all?
+    @problemReleased = @submission.scores.pluck(:released).all? &&
+                       !@assessment.before_grading_deadline?
```
Good catch! I would suggest making sure this is done in annotations_controller too, since I think the viewFeedback view also has a problemReleased check.

damianhxy replied:
Good point, verified that students are currently able to view annotations before grading deadline on the viewFeedback view.

michellexliu commented on app/assets/javascripts/annotations.js:
```diff
@@ -471,7 +529,75 @@ function newAnnotationFormCode() {
+    if (newAnnotation) {
+      submitNewAnnotation(comment, shared_comment, true, score, problem_id, 0, $(this));
+    } else {
+      const annotationObject = getAnnotationObject(config.annotationId);
```
nit: same thing as prev nit about newAnnotation object

damianhxy replied:
Updated to use `Object.assign` instead

michellexliu commented on docs/features/annotations.md:
```diff
@@ -2,28 +2,50 @@
 
-!!! attention "Non-Autograded Problems Only"
-    Note that annotations can only be added to non-autograded problems. Specifically, a problem is non-autograded if there is no assigned score for that problem in the json outputted by the autograder
+When the "Add to shared comment pool" option is checked, the comment will be saved to a shared comment pool.
```
might want to add info about who the shared comment pool is shared with and whether the pool is per-assignment vs per-problem.

damianhxy replied:
Done, updated docs to clarify that shared comment pool is shared with other instructors and course-assistants, and that the pool is per assessment.

michellexliu commented on app/views/submissions/_annotation_global.html.erb:
```diff
@@ -0,0 +1,4 @@
+<div class="base-global-annotation-form global-annotation-form">
```
May need to add a `display: none` to solve "Global annotation" header text flashing
![image](https://user-images.githubusercontent.com/50491000/204153693-1614ee8e-f6f3-46b1-bf89-2cf9655a82ca.png)

damianhxy replied:
Done

michellexliu commented on app/views/submissions/_annotation_global.html.erb:
```diff
@@ -0,0 +1,4 @@
+<div class="base-global-annotation-form global-annotation-form">
+  <h1> Global annotation </h1>
```
Nit: per accessibility best practices, only the main heading of a page should use an h1 tag; should use a lower-ranked heading element here

damianhxy replied:
I am using a `h1` tag because of the styling specific to `.annotationSummary h1`

<img width="310" alt="Screenshot 2022-11-27 at 14 52 41" src="https://user-images.githubusercontent.com/9074856/204156631-7889155b-51a1-4874-8133-f695ebf8e5ac.png">

Until recently, it was used for the annotations box 

![Screenshot 2022-11-27 at 14 54 07](https://user-images.githubusercontent.com/9074856/204156715-6f45d6fb-ecac-4043-8e90-30ddf84a3392.png)

In that case, should I update the style to use a `h2` tag or similar?

damianhxy replied:
Replaced with a `div`, and simplified the styles

michellexliu commented on app/views/submissions/_annotation_pane.html.erb:
```diff
@@ -3,48 +3,101 @@
+                  Sort global annotations by id (a[4])
+                  Sort file annotations by filename (a[6]), followed by line (a[2]) and group by filename
+              %>
+              <% annotations_by_type = descriptTuples.group_by { |a| a[7] }
```
would it be possible to move this logic to the controller?

damianhxy replied:
Makes sense

michellexliu commented:
Found a bug:

Create an annotation for a problem --> edit assessment & delete the problem --> error when you go to load the assignment again:
![image](https://user-images.githubusercontent.com/50491000/204310983-ac369237-95c7-4bd9-a331-f4f4e8eed9c7.png)

damianhxy commented:
> Found a bug:
> 
> Create an annotation for a problem --> edit assessment & delete the problem --> error when you go to load the assignment again: ![image](https://user-images.githubusercontent.com/50491000/204310983-ac369237-95c7-4bd9-a331-f4f4e8eed9c7.png)

What should be the expected behavior? In master, the current behavior seems to allow the annotations to persist and continue affecting the scores.

damianhxy commented:
> Found a bug:
> 
> Create an annotation for a problem --> edit assessment & delete the problem --> error when you go to load the assignment again: ![image](https://user-images.githubusercontent.com/50491000/204310983-ac369237-95c7-4bd9-a331-f4f4e8eed9c7.png)

Remedied by automatically destroying annotations when deleting a problem and updated the warning message when deleting a problem.

Just in case, I added code to handle any annotations that were somehow orphaned.

**Before deletion**
![Screenshot 2022-11-28 at 16 52 42](https://user-images.githubusercontent.com/9074856/204388800-f6114edc-7749-437d-9426-dcdf8c8fccb9.png)

**After deletion**
![Screenshot 2022-11-28 at 16 53 05](https://user-images.githubusercontent.com/9074856/204388856-d56429db-f62e-4df5-9de3-e65afa2d6755.png)

**New warning message**
![Screenshot 2022-11-28 at 16 53 29](https://user-images.githubusercontent.com/9074856/204388914-12973f4f-b043-4029-ac21-4f33af87ac58.png)

michellexliu commented:
lgtm! just fix the merge conflicts

victorhuangwq commented on app/assets/javascripts/annotations.js:
```diff
@@ -294,10 +295,13 @@ function displayAnnotations() {
 
   _.each(annotationsByPositionByLine[currentHeaderPos], function (arr_annotations, line) {
     _.each(arr_annotations, function (annotationObj, ind) {
+      if (annotationObj.global_comment === true)
```
Do we need an equality check here? would `annotationObj.global_comment` by itself work?

damianhxy replied:
Good point, that would work

victorhuangwq reviewed (approved):
LGTM. Small nit
